tf_quant_finance/black_scholes/approximations/american_option.py

```diff
@@ -0,0 +1,306 @@
+    max_iterations=max_iterations, tolerance=tolerance, name=name)
+
+  return tf.where(calculate_eu, eu_prices, am_prices), converged, failed
+
+
+def _option_price(*,
+                  sigma,
+                  x,
+                  t,
+                  r,
+                  b,
+                  s=None,
+                  is_call_options=None,
+                  dtype=None,
+                  max_iterations=20,
+                  tolerance=1e-8,
+                  name=None):
+  """Computes the price for a batch of call or put options, using an approximate
+    pricing formula, the Baron-Adesi Whaley approximation
+
+    The naming convention will align variables with the variables named in
+    reference [1], but made lower case, and differentiating between put and
+    call option values with the suffix _put and _call.
+
+    ## References:
+    [1] Baron-Adesi, Whaley, Efficient Analytic Approximation of American Option
+      Values, The Journal of Finance, Vol XLII, No. 2, June 1987;
+      https://deriscope.com/docs/Barone_Adesi_Whaley_1987.pdf
+
+    Returns:
+      option_prices: A `Tensor` of the same shape as `forwards` or `spots`.
+      The Baron-Adesi-Whaley approximation of the price of the american options.
+    """
+
+  with tf.name_scope(name):
+    where_call = tf.broadcast_to(
+      tf.constant(True), s.shape) if is_call_options is None else is_call_options
+
+    sign = tf.where(
+      where_call, tf.constant(1, dtype=dtype), tf.constant(-1, dtype=dtype))
+
+    q2, a2, s_crit, converged, failed = _option_price_formula_variables(
+      sigma, x, t, r, b, sign,
+      is_call_options, dtype, max_iterations=max_iterations, tolerance=tolerance)
+
+    eu_prices = vanilla_prices(
+      volatilities=sigma, strikes=x, expiries=t, spots=s, discount_rates=r,
+      cost_of_carries=b, is_call_options=is_call_options, dtype=dtype,
+      name=name)
+
+    # The divisive condition is different for put and call options
+    condition = tf.where(where_call, s < s_crit, s > s_crit)
+
+    american_prices = tf.where(
+      condition,
+      eu_prices + a2 * (s / s_crit) ** q2,
+      (s - x) * sign)
+
+    return american_prices, converged, failed
+
+
+def _option_price_formula_variables(sigma,
+                                    x,
+                                    t,
+                                    r,
+                                    b,
+                                    sign,
+                                    is_call_options,
+                                    dtype,
+                                    max_iterations=20,
+                                    tolerance=1e-8):
+  """ Calculates variables q, a and the critical spot price as per reference [1]
+  for the american option price formula.
+  These are values needed to calculate the approximate american option price.
+  For spot prices above and below the critical spot price, the american option
+  price formula behaves differently.
```

Comment:
Same as above

Reply:
i think this is done - let me know is this is what you were thinking (i thought this one in specific could use a bit of explanation so i've saved the old doc string in case you want some of if back :) )